### PR TITLE
[RFC] Add support for missing XHR response types

### LINKED ISSUE #6679: response.json() hangs when chrome debugging
When using fetch to make a get to an external api, and if debugging in chrome is active, then it never reaches the second "then" until there is an interaction with the UI, such as a tap on the screen, but if debugging is turned off then this problem goes away.

```
fetch(requestURL)
    .then((res) => res.json())
    .then((resData) => {
        this.setState({
            suggestions: resData.Similar.Results
        });
    })
    .done()
```

Using React Native 0.22
Only occurs on Android (both Genymotion and on device (5.0))
Using a Mac

### PR DESCRIPTION
**Pull request by davidaurelio**
Fixes #6679

This adds support for the missing response types to XMLHttpRequest.
Don’t ship this yet. This is completely untested. yolo and stuff.

### CODE REVIEW

**Comment by davidaurelio** [comment]
@vjeux @nicklockwood @javache  @ide @mkonicek @bestander @satya164

**Comment by davidaurelio** [comment]
@facebook-github-bot import

File: Libraries/Network/XMLHttpRequestBase.js [hunk 1]
```diff
@@ -86,20 +96,90 @@ class XMLHttpRequestBase {
...
+    }
+    if (!SUPPORTED_RESPONSE_TYPES.hasOwnProperty(responseType)) {
+      warning(
+        `The provided value '${responseType}' is not a valid 'responseType'.`);
+      return;
+    }
+
+    if (!SUPPORTED_RESPONSE_TYPES[responseType]) {
+      warning(
+        `The provided value '${responseType}' is unsupported in this environment.`);
```
Comment (davidaurelio) [inline Libraries/Network/XMLHttpRequestBase.js]: should this be an invariant? (except for `'document'` maybe)

**Comment by bestander** [comment]
Your lint neglect is shameful

**Comment by davidaurelio** [comment]
wanna take over?

File: Libraries/Network/XMLHttpRequestBase.js [hunk 2]
```diff
@@ -86,20 +96,90 @@ class XMLHttpRequestBase {
...
+        this._cachedResponse = null;
+        break;
+
+      case 'arraybuffer':
+        this._cachedResponse = toArrayBuffer(
+          this.responseText, this.getResponseHeader('content-type'));
+        break;
+
+      case 'blob':
+        this._cachedResponse = new global.Blob([this.responseText]);
```
Comment (ide) [inline Libraries/Network/XMLHttpRequestBase.js]: The Blob constructor accepts some options including the MIME type of the response -- should we parse the HTTP response headers and pass the Content-Type to the Blob too?

File: Libraries/Network/XMLHttpRequestBase.js [hunk 3]
```diff
@@ -86,20 +96,90 @@ class XMLHttpRequestBase {
...
     this._requestId = null;
 
+    this._cachedResponse = undefined;
+    this._hasError = false;
     this._headers = {};
+    this._responseType = '';
     this._sent = false;
     this._lowerCaseResponseHeaders = {};
 
     this._clearSubscriptions();
```
Comment (davidaurelio) [inline Libraries/Network/XMLHttpRequestBase.js]: ORLY? for `constructor`?

File: Libraries/Network/XMLHttpRequestBase.js [hunk 4]
```diff
@@ -86,20 +96,90 @@ class XMLHttpRequestBase {
...
   }
 
+  get responseType() {
+    return this._responseType;
```
Comment (davidaurelio) [inline Libraries/Network/XMLHttpRequestBase.js]: pity

File: Libraries/Network/XMLHttpRequestBase.js (hunk 2 above)
Comment (davidaurelio) [inline Libraries/Network/XMLHttpRequestBase.js]: I guess that makes sense

File: Libraries/Network/XMLHttpRequestBase.js (hunk 1 above)
Comment (ide) [inline Libraries/Network/XMLHttpRequestBase.js]: Yea, might be nice to get a clear error message early on.

File: Libraries/Network/XMLHttpRequestBase.js [hunk 5]
```diff
@@ -304,4 +370,52 @@ XMLHttpRequestBase.HEADERS_RECEIVED = HEADERS_RECEIVED;
...
 XMLHttpRequestBase.DONE = DONE;
 
+function toArrayBuffer(text, contentType) {
+  const {length} = text;
+  if (length === 0) {
+    return new ArrayBuffer();
+  }
+
+  /*eslint-disable no-bitwise, no-sparse-arrays*/
+  const charset = (contentType.match(/;\s*charset=([^;]*)/i) || [, 'utf-8'])[1];
```
Comment (ide) [inline Libraries/Network/XMLHttpRequestBase.js]: May want to trim this string

File: Libraries/Network/XMLHttpRequestBase.js [hunk 6]
```diff
@@ -304,4 +370,52 @@ XMLHttpRequestBase.HEADERS_RECEIVED = HEADERS_RECEIVED;
...
+
+  let array;
+  if (/^utf8$/i.test(charset)) {
```
Comment (ide) [inline Libraries/Network/XMLHttpRequestBase.js]: `/^utf-?8$/i`

File: Libraries/Network/XMLHttpRequestBase.js (hunk 5 above)
Comment (ide) [inline Libraries/Network/XMLHttpRequestBase.js]: also this code is kind of clever 😒

File: Libraries/Network/XMLHttpRequestBase.js (hunk 6 above)
Comment (davidaurelio) [inline Libraries/Network/XMLHttpRequestBase.js]: uh oh. good catch!

**Comment by ide** [comment]
Looks mostly good to me. I didn't look at the UTF-8 decoder too closely but assume it's right.

**Comment by davidaurelio** [comment]
Thanks for the review!

>  I didn't look at the UTF-8 decoder too closely but assume it's right.
it actually doesn’t handle surrogate pairs.

**Comment by davidaurelio** [comment]
@facebook-github-bot shipit

**Comment by davidaurelio** [comment]
@bestander, this is ready for merge. I assume you have to accept it in phabricator

**Comment by davidaurelio** [comment]
@facebook-github-bot import

**Comment by davidaurelio** [comment]
@ide, @bestander: as @javache assumed, `ArrayBuffer` / `Uint8Array` are unavailable on iOS7, even though they are activated in Safari.

What shall we do? Not support `xhr.responseType = 'arraybuffer'` at all? Setting it will throw an error on iOS7 right now. Android has support for it, I just checked the open source version

**Comment by bestander** [comment]
Do we even support ios7?
Otis like 2012

**Comment by ide** [comment]
Let's support `arraybuffer` responses on platforms that support it so that we don't hold back React Native for an OS with small market share. Apple says < 5% of users are on iOS 7.

This is the behavior I believe we want:
- Responses of type `arraybuffer` work on Android and iOS 8+
- React Native should not crash on iOS 7 if you never get an `arraybuffer` response
- If you do get an `arraybuffer` response, we should fail quickly and with a clear error

This should work for apps that want to target iOS 7 and you can easily write code like: `if (isIOS7) { dontFetchAnArrayBuffer(); }`. And for apps that are iOS 8+ you get the functionality you expect.

**Comment by ide** [comment]
@davidaurelio also could you add a comment to the code mentioning iOS 7 so that when we drop support for iOS 7 some day, it's easy to grep for deletable code?

File: Libraries/Utilities/utf8.js [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+  _realloc() {
+    const storage = this._storage;
+    this._storage = new Uint8Array(align(storage.length * 1.5));
+    this._storage.set(storage);
+  }
+}
+
+/*eslint-disable no-bitwise */
+exports.encode = (string: string): ArrayBuffer => {
```
Comment (javache) [inline Libraries/Utilities/utf8.js]: I might be missing something here, but doing utf8 ourself doesn't seem like something we should do.

Like the #1 rule of crypto "don't roll your own", I think this applies to string encoding too.

File: Libraries/Network/XMLHttpRequestBase.js [hunk 8]
```diff
@@ -13,14 +13,28 @@
 
 var RCTNetworking = require('RCTNetworking');
 var RCTDeviceEventEmitter = require('RCTDeviceEventEmitter');
-var invariant = require('fbjs/lib/invariant');
+const invariant = require('fbjs/lib/invariant');
+const utf8 = require('utf8');
```
Comment (vjeux) [inline Libraries/Network/XMLHttpRequestBase.js]: How big is this library?

Comment (vjeux) [inline Libraries/Network/XMLHttpRequestBase.js]: Nevermind -_- you implemented it below.

File: Libraries/Utilities/utf8.js (hunk 7 above)
Comment (davidaurelio) [inline Libraries/Utilities/utf8.js]: I think the complexity of crypto and character encodings is very different. I have been looking for a compact utf-8 encoding lib, but didn’t find anything that was compact *and* solid at the same time. My implementation adds ~1KB when minified.

Do you have any specific concerns? We wouldn’t need to do this if we could pass the binary response data from native to JS wrapped into a host object we’d implement. That host object could pass of character encoding to a system service. I have been looking into JSC’s API for that, but as long as we pass messages as JSON, we can’t do that.

**Comment by davidaurelio** [comment]
@facebook-github-bot import

**Comment by davidaurelio** [comment]
@facebook-github-bot import

**Comment by ide** [comment]
Cherry-picking this into 0.24-stable.

**Comment by mpretty-homepass** [comment]
@davidaurelio Just FYI the utf8 module this provides is supplanting a utf8 node module (https://github.com/mathiasbynens/utf8.js); I assume this is because of the '@providesModule' - should this really be doing that if all it does is encode?

**Comment by davidaurelio** [comment]
@mpretty-homepass sorry about that :-( we’ll get rid of the special packager rule for react-native.

Feel free to send a PR that renames the module to something else.

**Pull request closed without merging**